force-app/main/default/classes/SOQL.cls

```diff
@@ -288,20 +325,28 @@ public with sharing class SOQL {
      */
     public static List<Account> findContactsByLeadEmailDomains(Set<Id> leadIds) {
         // Query the Leads
-
+        Lead[] leadQuery = [SELECT Email FROM Lead WHERE Id = :leadIds WITH SECURITY_ENFORCED];
+        
         // Create a set of email domains from the Leads
-
+        Set<String> leadDomains = new Set<String>();
         // Loop through each Lead
-
-            // Get the email from the Lead
-
-            // Check if the email is not null
-
-                // Add the email domain to the set
-
+        // Get the email from the Lead
+        // Check if the email is not null
+        // Add the email domain to the set
+            for(Lead leads:leadQuery){
+                if(leads.Email != null){
+                    //Integer index = leads.email.indexOf('@');
+                    String leadEmailDomain = leads.Email.substring(leads.Email.indexOf('@')+ 1);
+                    String leadDomLike = '%'+leadEmailDomain+'%';
+                    leadDomains.add(leadDomLike);  
+                }
```

Comment:
can do a simple if (lead.Email == null) { continue; } statement, then do your logic.

great work with this problem.